File: hadesboot/src/main/kotlin/hadesc/ast/FunctionSignature.kt
package hadesc.ast

import hadesc.location.HasLocation
import hadesc.location.SourceLocation

data class FunctionSignature(
    override val location: SourceLocation,
    override val defId: DefId,
    val name: Binder,
    override val typeParams: List<TypeParam>?,
    val thisParamBinder: Binder?,
    val thisParamFlags: ThisParamFlags?,
    val params: List<Param>,
    val returnType: TypeAnnotation,
    val whereClause: WhereClause?
) : HasLocation, HasTypeParams, HasDefId {
    init {
        check(
            thisParamBinder == null && thisParamFlags == null ||
                thisParamBinder != null && thisParamFlags != null
        )
    }
    data class ThisParamFlags(
        val isPointer: Boolean,
        val isMutable: Boolean
    )
}

File: lib/ast/FunctionSignature.cpp
//
// Created by dhruv on 07/08/20.
//
#include "hades/ast/FunctionSignature.h"

namespace hades {

FunctionSignature::FunctionSignature(   //
    SourceLocation location,            //
    Identifier name,                    //
    FunctionSignature::Params &&params, //
    Optional<const TypeAnnotation *> return_type  //
    ) noexcept                          //
    : m_location(location),             //
      m_name(name),                     //
      m_params(params),                 //
      m_return_type_annotation(return_type)        //
{}

auto FunctionSignature::location() const -> const SourceLocation & {
  return m_location;
}
auto FunctionSignature::params() const -> ArrayRef<const Param *> {
  return hades::ArrayRef<const Param *>(m_params);
}

auto FunctionSignature::return_type_annotation() const -> const Optional<const TypeAnnotation *> & {
  return m_return_type_annotation;
}

auto FunctionSignature::name() const -> const Identifier & {
  return m_name;
}

} // namespace hades
